articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+- Vector: This is SSE or AVX datatype (we won't talk about this type-kind).
+- Aggregate: General type-kind for structs and classes (we will talk about this type-kind in the [structs](#structs) section).
+
+Sidebar: Booleans in [C](https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf#section.3.2), [C++](https://isocpp.org/files/papers/n4296.pdf#section.3.9), and [Rust](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#bool) all *only* have acceptable values of 0 and 1 (even though the size is 1 byte or 8 bits). This is because, as shown in these examples, [C](https://repl.it/@ZackJorquera/C-bools-only-0-or-1), [C++](https://repl.it/@ZackJorquera/Cpp-bools-only-0-or-1), and [Rust](https://repl.it/@ZackJorquera/Rust-bools-only-0-or-1) only look at the 0th bit to determine the truth value. This means the while a value like `4`, `0b0100`, is not zero it will still be false as a bool (technically it is undefined behavior). However, in an if statement and when casting there is an [implicit conversion](https://isocpp.org/files/papers/n4296.pdf#section.4.12) that will take any non-zero value and make it have the value 1 if it isn't already the `bool` type.
+
+We will talk more about these type-kinds in the [structs](#structs) and [calling conventions](#passing-data-to-the-callee-and-back) sections.
+
+When it comes to primitive types the main issues to think about are how to store them in memory and how to operate on them. These two things play hand in hand because if we want to operate on some data we will first need to read them from memory and into a register.
+
+For anything stored in memory, we will have to think about the [endianness](https://www.section.io/engineering-education/what-is-little-endian-and-big-endian/) of the data. Most of the time this is dealt with by the hardware and exposed to the program as a `mov` instructions that already account for endianness with exception to loading constant value with `mov`. I won't go into detail as I wrote a whole article about it [here](https://www.section.io/engineering-education/what-is-little-endian-and-big-endian/).
+
+## Size and Alignment
+
+There are also the size and alignment of each piece of data that we have to worry about. The size is, well the size but must also be a multiple of the alignment. The alignment specifies where it can be stored in memory, it must be at least 1 and always a power of two. As an example, if something has an alignment of 4-bytes then the address it is stored at must be divisible by 4-bytes. The reason why this exists is mostly for [performance reasons](https://stackoverflow.com/q/49391001/9664285) because memory hardware is specialized in such a way to read data from cache-lines and alignment can guarantee that one piece of data is not split up into multiple cache-lines.

```

Comment:
There are also -> There is also

Reply:
Fix this sentence: The size is, well the size but must also be a multiple of the alignment.

Reply:
As an example, if something has an alignment -> As an example, if some block of memory has an alignment

Reply:
The reason why this exists is mostly -> This exists  mostly

Reply:
alignment can guarantee that -> alignment and can guarantee that

Reply:
I dont this your changes work. This is what I changed it to:
There are also the size and alignment of each piece of data that we have to _worry about (this follows for structs as well). The size must always be a multiple of the alignment._ The alignment specifies where it can be stored in memory, it must be at least 1 and always a power of two. _As an example, if something has an alignment of 4-bytes then the address it is stored at must be divisible by 4-bytes (and the size must be divisible by 4-bytes)._ __This exists mostly for [performance reasons](https://stackoverflow.com/q/49391001/9664285) because memory hardware is specialized in such a way as to read data from a cache-block (also called a cache-line) quickly. The alignment can guarantee that one piece of data is not split up into multiple cache-blocks which would require two memory reads.__

Reply:
There are also the -> There is also the

Reply:
memory, it must be -> memory; it must be

Reply:
alignment of 4-bytes then the -> alignment of 4-bytes, then the

Reply:
> There are also the -> There is also the

Maybe I just done understand grammar.